### DISCUSSION TITLE
What is the way to combine multiple failures when T is different but E is the same?

### DISCUSSION THREAD

**Discussion by xavierjohn** [vkhorikov/CSharpFunctionalExtensions#377] [Q&A] [answered]
I am trying to understand the methods in Result.Combine (CSharpFunctionalExtensions\Result\Methods\Combine.cs)

I have an Error class so the Result I use is in the form Result<T,Error>.

Example:
```
var rFirstName = FirstName.Create(strFName);
var rLastName = LastName.Create(strLName);
var rEmail = EmailAddress.Create(strEmail);
var result = Result.Combine(rFirstName, rLastName, rEmail);
```

Assuming the Error type is Error for all the Create methods, how do I get a collection of Errors? In the tests I saw T is the same but in my case T is different, only Error is the same type.

Thanks.

**Comment by xavierjohn** [comment, marked as answer]
This is my solution and hopefully there is a better way.

I created an ErrorCollection class with an implicit operator to convert error.
```
    public sealed class ErrorCollection : ICombine
    {
        public ErrorCollection() => _errors = new List<Error>();
        public ErrorCollection(IEnumerable<Error> errors) => _errors = errors.ToList();
        public Error[] Errors { get => _errors.ToArray(); }
        public bool HasErrors => _errors.Any();

        public void Add(ErrorCollection ec) => _errors.AddRange(ec.Errors);

        private readonly List<Error> _errors;

        ICombine ICombine.Combine(ICombine value)
        {
            var errorCollection = (ErrorCollection)value;
            var errorList = new List<Error>(errorCollection.Errors);
            errorList.AddRange(Errors);
            return new ErrorCollection(errorList);
        }

        public static implicit operator ErrorCollection(Error e) => new ErrorCollection(new List<Error>() { e });
    }
```

Then I can combine the errors as
`var result = Result.Combine<ErrorCollection>(rFirstName , rLastName, rEmail );`

**Comment by vkhorikov** [comment]
Yes, such an `ErrorCollection` is what I meant in the previous post when I said that the library should support combining out of the box. This needs to be in the library (though I suggest to title it `ErrorList` just because it's more succinct). Then, we can make `Result.Combine` return `ErrorList` instead of just single error.

**Reply by xavierjohn** [reply]
I have made a pull request
https://github.com/vkhorikov/CSharpFunctionalExtensions/pull/378
